support custom queue for the scheduler

suigh opened the pull request:
docs for KEP issue #2779

denkensk commented:
/cc

denkensk commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+	command := app.NewSchedulerCommand(
+		app.WithPlugin(coscheduling.Name, coscheduling.New),
+                        …...
+	+	app.WithCustomQueue(myQueue.New),
```
I think It maybe  better to make the CustomQueue as a plugin and register it as other plugins like:
```go
	command := app.NewSchedulerCommand(
		app.WithPlugin(coscheduling.Name, coscheduling.New),
	+	app.WithPlugin(myQueue.Name, myQueue.New),
)

suigh replied:
Thanks for the suggestion, please see my answer below, it explains why `WithPlugin` is not re-used.

denkensk reviewed (commented):
It's a particularly good idea.

Some suggestion:
1. It may be better to make the queue as the new extension point of scheduling framework. Then the users can implement their custom queue as the plugins and register or construct like other plugins.
2. Can you give a brief description of the interface to be implemented in custom queue?

suigh commented:
@denkensk 
Thanks for your suggestions, please see my answers here

`It may be better to make the queue as the new extension point of scheduling framework. Then the users can implement their custom queue as the plugins and register or construct like other plugins.`
My first idea is like this one, and I switched to the current one as 
1. we store the registered plugin in the following struct
type Registry struct {
	map[string]PluginFactory
}
type PluginFactory = func(configuration runtime.Object, f framework.Handle) (framework.Plugin, error)

`PluginFactory ` is not the expected function to create a custom queue as the queue needn't the parameters in it

2. With this solution, we will have to identify the input's type  -- for plugin or for queue somewhere, and then assign the custom queue to the scheduler. This check is not straightforward, add a new function like `WithCustomQueue` seems better.


`Can you give a brief description of the interface to be implemented in custom queue?`
User will implement the interface `SchedulingQueue` for the current queue. User can just copy the `pkg\scheduler\queue\scheduling_queue.go` and update something. I will add an example into the  SIG_scheduling after updating the k8s codes.

`Some concerns:
It maybe difficult for users to implement a complete custom queue. Is it considered that only part of the interface is exposed like pop or add, or the data structure like heap?`
Considered this before, and choose the current one because:
1. by this way, the solution will be complex than the current one, and when user ask for more functions, we will have to update the codes again.
2. for an user who only want to update one or several functions, the user can just copy the `scheduling_queue.go` and update it, and left other functions there., the effort is same as this way.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+## Summary
+
+Provide an interface so that user can implement a custom queue of scheduler,
```
```suggestion
Provide an interface so that users can implement a custom queue of scheduler.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+## Summary
+
+Provide an interface so that user can implement a custom queue of scheduler,
+with such a queue, user can sort the pods in the queue with specified rules 
+and get some functions that are not supported now.
```
```suggestion
With such, users can build their custom queue implementation at will, and plumb into
the scheduler seamlessly. This can satisfy the business needs such as sophisticated
pods sorting and internal queuing mechanics.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+and get some functions that are not supported now.
+
+## Motivation
+The current queue (PriorityQueue in pkg\scheduler\internal\queue\scheduling_queue.go) 
```
```suggestion
The current internal queue implementation ([scheduling_queue.go#PriorityQueue](https://github.com/kubernetes/kubernetes/blob/ee459b8969ed2abfed79a07d4ac9d41f13f18ce6/pkg/scheduler/internal/queue/scheduling_queue.go#L126)).
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+## Motivation
+The current queue (PriorityQueue in pkg\scheduler\internal\queue\scheduling_queue.go) 
+for pod selection in the scheduler is good enough for most cases, while it cannot handle 
+all the requests even with a custom scheduler based on the kubernetes scheduler plugins.
```
```suggestion
works for most cases, but the limited interfaces exposure makes it hard to
implement a scheduler plugin for sophisticated requirements.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+for pod selection in the scheduler is good enough for most cases, while it cannot handle 
+all the requests even with a custom scheduler based on the kubernetes scheduler plugins.
+
+One example:
```
```suggestion
One requirement in terms of multi-tenancy support:
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+One example:
+There are many pods from 2 users (userA and userB), and the target is to ensure resource
+usage (E.g. userA gets x CPUs) ratio between userA and userB is 1:1. Sorting the pods by
+the resource usage in Less function cannot work, because the resource usage for 
```
```suggestion
the resource usage in `Less` function doesn't work properly, because the resource usage for 
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+algorithm (heapsort) that used by the current queue to get the next pods depends on static
+data.
+
+Another example:
+The interval of function flushUnschedulableQLeftover is hard coded to 30 seconds, user
+cannot change it.
```
It's now customizable, so no longer a valid motivation.

Huang-Wei replied:
So, can you compile at least one more motivation here?

suigh replied:
Update the message to:

Another requirement in terms of custom function support:  
For example, the interval of function flushUnschedulableQLeftover cannot be updated dynamically, while
user want it to be shorter or longer at different times as shorter interval can get pods to be scheduled
faster and longer one can get less logs flushed.

Huang-Wei replied:
Yeah, that's a fair point.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+The interval of function flushUnschedulableQLeftover is hard coded to 30 seconds, user
+cannot change it.
+
+These issues can be resolved if user can implement the queue by themselves, then user
+can get what they want easily, and code maintainer needn't handle various requests.
```
```suggestion
Given that the business needs may vary greatly, it'd be good to provide a replaceable
`SchedulerQueue` interface and plumbing mechanics. So that the developers can
implement their custom queue implementation, while the upstream maintainers focus
on keeping the core small and extensible, and thus only maintain the internal queue piece.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+### Goals
+
+Support custom queue of the scheduler.
```
Support custom queue implementation of the scheduler

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+### Non-Goals
+
+Support custom rules for pod selection in the current queue of scheduler
```
```suggestion
Support custom rules for pod selection in the current internal queue of scheduler.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+Provide an interface like current scheduler plugins, users can provide a
+custom queue of scheduler and register it, then the kubernetes
+scheduler will use this custom queue for pod selection.
```
```suggestion
scheduler will use this custom queue for pod management.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+more details of the scheuler with this enhancement.
+
+Pros:  
+Users can get full control of the pod selection logic in the queue, they
```
```suggestion
Users can get full control of the pod queuing logic, they
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+Pros:  
+Users can get full control of the pod selection logic in the queue, they
+can select pods with custom logic.
```
```suggestion
can pop/re-queue/backoff pods with custom logic.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,423 @@
+
+### Risks and Mitigations
+
+If custom queue is not enabled or provided, the current queue will work as
+before, no other users will be impacted.
```
A poor-implemented queue may not function well, in both functionality and performance.
But the default scheduler works as before and thus won't be impacted.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,423 @@
+  ```
+	command := app.NewSchedulerCommand(
+		app.WithPlugin(coscheduling.Name, coscheduling.New),
+                        …...
````
```suggestion
                        ......
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,423 @@
+1. User implements a custom queue with the interface SchedulingQueue,
+the name is myQueue, and registers it
+
+  ```
````
```suggestion
```go
```

suigh replied:
I removed the space. Why adding `go`? A typo?

Huang-Wei replied:
that's for pretty formatting.

suigh replied:
Thanks for this suggestion. 
Downloaded an extention for markdownlint to ensure the markdown format can be followed.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,423 @@
+   ```
+
+### Test Plan
+
````
Please refer to other KEPs to fill this section as well as others.

suigh replied:
Updated with 

```
- **Unit Tests**: All core changes must be covered by unit tests.
- **Integration Tests**: One integration test for the custom queue.
- **Benchmark Tests**: The performance benchmark test result is same as before if custom queue is not used.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,423 @@
+      podQueue = schedulingQueue.NewSchedulingQueue()
+  }
+   ```
+
````
Can you also add here to explain how it explains with other internal scheduler structures like Cache, Assumed state, NominatedPods, etc. Any other interfaces are needed? and how. What's the risk here?

suigh replied:
A new interface Init will be aded to handle the initialization parameters from the k8s framework, it is used to replace the old function NewPriorityQueue, the logic inside them are same.
Packages queue and heap will be exposed to end users, then users can access/use them and the mentioned internal scheduler structures. 


```
4. Basic inputs for queue's initialization will not be changed, a new interface `Init` is added in
SchedulingQueue, users must implement it.
-func NewPriorityQueue(
+func (p *PriorityQueue) Init(
 	lessFn framework.LessFunc,
 	informerFactory informers.SharedInformerFactory,
 	opts ...Option,
)
5. For the scheduler internal packages under ([internal] (https://github.com/kubernetes/kubernetes/tree/b6c75bee15e150628fcc240ab32dba6190d254e4/pkg/scheduler/internal)), queue and heap will be moved out of this folder so that users can access and use the structures inside them, cache and parallelize will not be touched.
```

Huang-Wei replied:
> Packages queue and heap will be exposed to end users, then users can access/use them and the mentioned internal scheduler structures.

I don't think I like this idea. The internal structure should be restricted for internal usage only.

Can you list the blockers (due to missing plumbing mechanics) that a pure out-of-tree implementation won't work properly?

Also, I'm also thinking about how `Option` can be more general and thus more usable for an out-of-tree queue implementation. Right now the internal `priorityQueueOptions` are fixed data structures, and hence not extensible.

suigh replied:
OK, let's keep queue and heap in the original location.  Removed `5` from the document.

The current blockers in my mind are `SchedulingQueue` and `PriorityQueueOptions` (maybe there will be more, I will write some codes to verify it), and I'm planning to move them to `pkg\scheduler\framework\interface.go` so that the users can access them, this is a right loaction for you?

suigh replied:
close this thread as there is same thread below.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.22"
+  beta: "v1.23"
+  stable: "v1.24"
```
We won't be able to finish in alpha in this release:

```suggestion
  alpha: "v1.23"
  beta: "v1.24"
  stable: "v1.26"
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+
+# The following PRR answers are required at beta release
+metrics:
+  - my_feature_metric
```
We can follow up with PRR later.

Huang-Wei replied:
Please don't fold leave unresolved comments.

suigh replied:
Oh, sorry. I misunderstand this comment, I think there will be another round review at `beta release` ....

alculquicondor replied:
Say TDB instead.

suigh replied:
updated.

Huang-Wei reviewed (commented):
Thanks @suigh. Here is the first round of review.

suigh commented:
@Huang-Wei 
Thanks very much for the detailed review, I had updated the documents, please help review the new versions.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -119,52 +124,52 @@ The codes will be updated as following:
-No need to outline all of the test cases, just the general strategy. Anything
-that would count as tricky in the implementation, and anything particularly
-challenging to test, should be called out.
+5. For the scheduler internal packages under ([internal] (https://github.com/kubernetes/kubernetes/tree/b6c75bee15e150628fcc240ab32dba6190d254e4/pkg/scheduler/internal)), queue and heap will be moved out of this folder so that users can access and use the structures inside them, cache and parallelize will not be touched.
```
remove the blank between text and URL link.

Again, I don't think it's feasible to move queue and heap out.

suigh replied:
Item `5` was removed, queue and heap will be internal packages as before.

suigh replied:
The current blockers in my mind are SchedulingQueue and PriorityQueueOptions (maybe there will be more, I will write some codes to verify it), and I'm planning to move them to pkg\scheduler\framework\interface.go so that the users can access them, this is a right loaction for you?

suigh replied:
Checked with codes, following items in the current queue need to be exposed, and I'm planning to move them into `pkg\scheduler\framework\interface.go`

```
type PreEnqueueCheck func(pod *v1.Pod) bool

type PriorityQueueOptions struct {}

type Option func(*PriorityQueueOptions)

type SchedulingQueue interface {}
```

suigh commented:
@Huang-Wei 
Updated doc as per the discussion in section `Design Details`, please help review. :)

What is updated:
1: for the coupled items (e.g., cache, option) with internal queue, this solution will not decouple them, we will just expose the coupled ones that need to be accessed by the custom queue. This can make the solution simple, and we can refine the codes in the future if necessary.
2: to get a custom queue, an user can copy the `scheduling_queue.go` and update it, then register to the scheduler, this is easy and straightforward.
3: update some previous messages to make the logic clear.

suigh commented:
some update:
1. document was updated at `Design Details` section, added more details about what will do and what will not do. The current design will not provide perfect interface as it is complex and we already can get what we want by the current design.

2. tried this solution by implementing a custom queue ( get a file named scheduling_queue.go, implemented similiar logic in scheduling_queue.go) and registered it like other plugins, the result is OK, pods were scheduled by the logic in customer quque, this solution works.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+
+## Design Details
+
+The current internal queue is coupled with `cache`, `Option` and other implementations
```
The scheduler _internal_ queue, as its name implies, was originally implemented for internal usage. Due to that, some structs (like `cache`, `Option`) are not designed for external extension, and the internal queue is initialized with specified parameters like `LessFunc`, `SharedInformerFactory`,
and `Option` by [factory.go#create](https://github.com/kubernetes/kubernetes/blob/f1f0183d2bbcde33024b2a05d6f39df32f11e037/pkg/scheduler/factory.go#L172)). Although refactoring the internal queuing mechanics to be reusable/extensible is the eventual goal, we'd like to approach that gradually, and thus is not the goal of phase 1 of this design. In phase 1, we'd like to exercise the idea of "a replaceable scheduler queue" in a "minimal viable product" manner.

suigh replied:
updated.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+### Goals
+
+Support custom queue implementation of the scheduler.
+
```
Add a bullet: Roll out the design gradually. In phase 1, keep current internal queue implementation intact.

suigh replied:
added.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+and then pass the custom queue to the scheduler like the plugins. With this way, user's
+logics will override the logics of the current internal queue.
+
+The codes will be updated as following:
```
A practical design for phase 1 is described as below:

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+
+The codes will be updated as following:
+
+1. Make the interface `SchedulingQueue` as a public one, so that users can implement it
```
```suggestion
1. Make `SchedulingQueue` a public interface. Users can choose to implement their own queue.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+1. Make the interface `SchedulingQueue` as a public one, so that users can implement it
+for the custom queue.
+
+2. Provide a new function `WithCustomQueue`, users can register the custom queue with this
```
```suggestion
2. Provide a new function `WithCustomQueue`, so users can register the custom queue with this
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+this solution will not decouple them as this has no impact on the goal of custom queue,
+and such logics can be enhanced in the future on demand.
+
+Some of the items like `SchedulingQueue`, `Option` that are only used by the internal
+queue now will be exposed at [interface.go](https://github.com/kubernetes/kubernetes/blob/fb9cafc99be94a73d9b92545164dbf336bbd230a/pkg/scheduler/framework/interface.go),
+users of custom queue can access them.
+
+To implement a custom queue, what an user needs to do is to copy the codes of the current
+internal queue (e.g., scheduling_queue.go), update some logics to get the custom queue,
+and then pass the custom queue to the scheduler like the plugins. With this way, user's
+logics will override the logics of the current internal queue.
```
Remove these 2 paragraphs as they are custom implementation details (or suggestions).

suigh replied:
removed.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,449 @@
+      +        opts ...framework.Option,
+      + ) {
+    ```
+
````
Alternatively, you can ass a stanza about the implementation practice/suggestion here:

```
To get started with the implementation, an easy way is to copy the codes of the current
internal queue (i.e., scheduling_queue.go), and then update the logic in-place to adapt to
your business needs  (it's very like how some out-of-tree PostFilter plugin refers to the defaultpreemption plugin).
With the custom queue handle, you can pass it to the scheduler like the plugins.
Following this way, the user's custom scheduler queue will override the upstream internal queue.
```

suigh replied:
added.

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+### Test Plan
+
+- **Unit Tests**: All core changes must be covered by unit tests.
+- **Integration Tests**: One integration test for the custom queue.
```
```suggestion
- **Integration Tests**: At least one integration test to craft a custom queue to exercise an end-to-end flow.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+- **Benchmark Tests**: The performance benchmark test result is same as before if custom queue is not used.
+
+### Graduation Criteria
+
```
Describe the plan of how to roll it out here. Like phase 1, phase 2, etc.

suigh replied:
This document is for the solution of phase 1 (provide interface to implement custom queue), and there is no solution details for phase 2 (refine the codes, decouple the queue with some structures), should we mention the `Graduation Criteria` of phase 2 in this document?

alculquicondor replied:
I think you should have *some* idea of what phase 2 would look like before we proceed.

suigh replied:
For phase 2, what we will do is
1. refine the codes, make the interface simple and easy to use, decouple this feature with other parts of the scheduler
2. refine some designs with user's feedback

Update the message as following:
```
 #### Beta -> GA Graduation

+- Only necessary interface and structs are exposed to users.
 - Allowing time for feedback to ensure that the new interface sufficiently expresses users requirements.
```

suigh replied:
For phase 2, what we will do is

refine the codes, make the interface simple and easy to use, decouple this feature with other parts of the scheduler
refine some designs with user's feedback
Update the message as following:

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+
+#### Alpha
+
+- Feature implemented behind a feature flag
```
I guess we don't need a feature gate here. But we can mention the maturity level by using the terms alpha, beta, etc.

suigh replied:
Updated as following, is this OK?
```
### Graduation Criteria

#### Alpha -> Beta Graduation

- Users can implement their own custom queues.
- No user complaints regarding correctness.

#### Beta -> GA Graduation

- Allowing time for feedback to ensure that the new interface sufficiently expresses users requirements.
```

Huang-Wei commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,449 @@
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
```
Polish this as well as other uncompleted ones.

suigh replied:
Removed this part, and add some details about the `Graduation Criteria`, mentioned by above answers.

Huang-Wei commented:
/cc @ahg-g @alculquicondor @damemi for awareness.

Huang-Wei commented:
@suigh LGTM generally. Please leave some time for other reviewers to take a look.

alculquicondor commented:
I have this in my queue. I will take a look later this week or next week.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+creation-date: 2021-06-09
+reviewers:
+  - "@Huang-Wei"
+  - TBD
```
Please add me

suigh replied:
added

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.22"
```
1.23

suigh replied:
updated to 1.23

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/kep.yaml:
```diff
@@ -0,0 +1,42 @@
+# The following PRR answers are required at alpha release
+# List the feature gate name and the components for which it must be enabled
+feature-gates:
+  - name: support custom queue for the scheduler
```
I suppose we don't need a feature gate for this feature, as it wouldn't change behavior.

suigh replied:
comment out this part as suggetsed.

alculquicondor commented on keps/prod-readiness/sig-scheduling/2779.yaml:
```diff
@@ -0,0 +1,3 @@
+kep-number: 2779
+beta:
```
alpha

suigh replied:
updated.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+### Goals
+
+1. Support custom queue implementation of the scheduler.
+2. Roll out the design gradually. In phase 1, keep current internal queue implementation
```
Please clarify: is it a goal to change the default queueing plugin? If not, please add a Non Goal. Regardless, point 2 needs is not very clear on the intent.

suigh replied:
This feature is only for the [scheduling_queue.go](https://github.com/kubernetes/kubernetes/blob/ee459b8969ed2abfed79a07d4ac9d41f13f18ce6/pkg/scheduler/internal/queue/scheduling_queue.go), no plugin will be touched. Do we need to plugin here?

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+
+## Summary
+
+Provide an interface so that users can implement a custom queue of scheduler.
```
which kind of user? end user or custom scheduler developers?

suigh replied:
custom scheduler developers, updated this part to show it explicitly.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+## Summary
+
+Provide an interface so that users can implement a custom queue of scheduler.
+With such, users can build their custom queue implementation at will, and plumb into
```
which scheduler? the default scheduler? IIUC, the intent is to easy the development of custom schedulers.

The phrasing of this paragraph makes it seem that you are proposing an API to configure the queue at runtime, which is not the case.

suigh replied:
The developers will get a scheduler binary with the custom queue, then they can replace the default scheduler binary with it, or run it as the secondary scheduler of the k8s.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+
+Other requirements come from current semi-exposed functions/parameters:
+For example, the interval of function flushUnschedulableQLeftover, although configurable,
+is immutable after it gets initialized. A requirement is to adjust the interval dynamically
```
What is the use case for this?

The intend of the flush is to make sure we are not leaving pods without scheduling retries. Ideally, it shouldn't exist and retries should entirely be based on cluster events. But realistically, it's hard to catch all events.
But depending on your use case, you might just want a mechanism to trigger the retry of *certain* pods.

suigh replied:
This case is used to show that there are many possible customer cases that we cannot handle with the current internal queue, and we need a solution for customers to implement the queue at their will. This is also the reason that why we want the customer the implement the whole queue.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+## Proposal
+
+Provide an interface like current scheduler plugins, users can provide a
+custom queue of scheduler and register it, then the kubernetes
```
at which time? build time, configuration time, runtime?

suigh replied:
At build time, the custom queue will be part of the scheduler binary after building.  Readme updated.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+custom logic.
+
+The scheduler plugins design is enhanced, the custom queue and the extension
+points can work together to meet more requests.
```
s/requests/requirements

suigh replied:
a typo?

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+One requirement in terms of multi-tenancy support:  
+There are many pods from 2 users (userA and userB), and the target is to ensure resource
+usage (E.g. userA gets x CPUs) ratio between userA and userB is 1:1. Sorting the pods by
+the resource usage in `Less` function doesn't work properly, because the resource usage for
```
Can you share a sample design of how you would implement a queue that satisfies this requirement?

Perhaps we can put the interface in a different place so that you don't have to implement the entire queue. I would like to see Alternative proposals.

suigh replied:
To resolve the issue that mentioned by me, I will build some sub-queues of the queue, one sub-queue for one user. Once schedule wants to pop a pod from the queue, the queue will go through the sub-queues, select the most under-used, and pop a pod from it. 

The purpose of this solution is to expose the entire queue, so that scheduler developer can control all the details in a queue, and we needn't expose any new function/structs/etc in the future. This is used to save our effort. :)

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,398 @@
+         )
+    ```
+
+3. The registered custom queue will be recorded by updated `Registry` struct.
````
This sounds like implementation detail that can be left for code reviews. I would probably ask to not touch the registry. But again, we don't need to discuss this in the KEP.

suigh replied:
Merged items 3/4 into one item, remove the code details, the new message is:

3. The registered custom queue will be passed to the scheduler if it exists, or the current internal queue will be used.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,398 @@
+      }
+    ```
+
+4. In function `(c *Configurator) create()`, custom queue will be passed to the scheduler
````
Also irrelevant for the KEP

suigh replied:
As mentioned above, merged it with item 3 and removed the code details.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -0,0 +1,398 @@
+
+A practical design for phase 1 is described as below:
+
+1. Make `SchedulingQueue` a public interface. Users can choose to implement their own queue.
```
The interface definition should be part of the KEP

suigh replied:
There is no change for the interface `SchedulingQueue`, just make it from private to publish, this is why its details are not listed. To avoid confusing, provide following message for more details:

1. Make `SchedulingQueue` a public interface which is private now in ([scheduling_queue.go](https://github.com/kubernetes/kubernetes/blob/ee459b8969ed2abfed79a07d4ac9d41f13f18ce6/pkg/scheduler/internal/queue/scheduling_queue.go#L126)). Users can choose to implement their own queue.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
````diff
@@ -0,0 +1,398 @@
+      + ) {
+    ```
+
+To get started with the implementation, an easy way is to copy the codes of the current
````
Irrelevant for the KEP

suigh replied:
this part was removed.

alculquicondor reviewed (commented):
As part of the alternatives, consider this:

(I'm assuming your use case is around batch, not serving applications).
Have you considered using a higher level controller and the Job.suspend feature to govern resource sharing? The idea is that you would suspend Jobs at creation. The Job would only be un-suspended (resumed) when the Pods for the Job wouldn't violate fairness or resource quotas or whatever policy you want to implement https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job

This solution is good for 2 reasons:
- There are no unschedulable Pods competing for scheduling time.
- The resume policy doesn't have to be based on the current size of the cluster, which makes sense when you have a cluster autoscaler.

denkensk commented:
> Have you considered using a higher level controller and the Job.suspend feature to govern resource sharing? The idea is that you would suspend Jobs at creation. The Job would only be un-suspended (resumed) when the Pods for the Job wouldn't violate fairness or resource quotas or whatever policy you want to implement

---

I like the idea. In fact, we have implemented higher level controller and used it in the production environment in Alibaba. We are preparing to open source the design and code. I'll invite you to help review. But this does not conflict with this kep. We pay more attention to the job level like tf-job mpi-job and so on. Maybe we can cooperate to solve some batch scheduling problems @Huang-Wei @alculquicondor @suigh

alculquicondor commented:
I would like tf-operator (and every batch API) to use k8s Jobs underneath so it's possible to suspend them.
See kubeflow/tf-operator#1303 for the ongoing discussion.

alculquicondor commented:
mpi-operator might be able to migrate to k8s Jobs sooner kubeflow/mpi-operator#373

denkensk commented:
Thanks for your information @alculquicondor 
We chose a similar way to express suspend by adding annotation here
https://github.com/kube-queue/api/blob/bffe1acb5aa961be9f171ff0fa2afa47f02f5381/pkg/apis/scheduling/v1alpha1/type.go#L99
And we also try to add the related processing logic to TF-operator and MPI-operator  @zw0610
But if you can change every batch API in the ecosystem) to use k8s Jobs underneath，It will bring us a big help

alculquicondor commented:
So back to the proposal in this KEP. I'm not convinced that the scheduling queue sorting is gonna guarantee any type of resource sharing, like this requirement listed in the motivation:

> There are many pods from 2 users (userA and userB), and the target is to ensure resource usage (E.g. userA gets x CPUs) ratio between userA and userB is 1:1.

The scheduling queue sorting is kind of like a heuristic about which pod should be tried next. But let's say you are able to implement the sorting based on users' usage ratio. We try to schedule Pod A and suddenly the API server was overwhelmed and couldn't accept the binding operation. Then Pod A is added back to the queue. But since binding happens in a separate routine, maybe we are already scheduling Pod B from another user. Then Pod B succeeds and you were not able to guarantee the ratio.

Then, what I think we should be doing is not care too much about the queue sorting. But then we can have a PreFilter (and perhaps a Permit) plugin that would return `Unschedulable` if such pod would violate the resource sharing.

If what I'm proposing doesn't make sense, I would like to see why in the "Alternatives" section of the KEP.

suigh commented:
@alculquicondor 
Thanks very much for your review and suggestions. 

For the solution of suspending a job, it works for sharing resources at job level, but it doesn't work for pod level sharing. E.g., if we can govern the resource sharing by stopping half of running pods in a job (the stopped pods will be re-created later), suspending the whole job is overkilled. 

Agree that `I'm not convinced that the scheduling queue sorting is gonna guarantee any type of resource sharing...`, only queue sorting cannot resolve this issue, we also need preemtption to preempt the pods from over-used users, then the pods from under-used users can run, and the KEP is to ensure the pods from under-used user can run first. Without this KEP, ping-pong preemption may happen as the pods from over-used users can be preempted and then get resource before other pods, and this step can be repeated.

For suggestion `But then we can have a PreFilter (and perhaps a Permit) plugin that would return Unschedulable if such pod would violate the resource sharing.`, I tried this one before, it is a little complex for developers, expecially when there are thousands of pods with different requests from differnt users, it is hard to say if a pod should run or not. For example, podA is from an over-used user, running it will violate the resource sharing, should we just mark it as `Unschedulable`? It is not true as if no other pods can run at all, podA should run to make full use of the cluster resources. It is hard to know the status of all other pods, even we can know, the performance will be impacted by going through all the pods, so it is hard to mark it as `Unschedulable`. With this KEP, we can control the sorting logic and always try the pods from under-used users first, and try other pods later, the above issue will not happen, the resources in the cluster can be fully used.

With this KEP and the scheduler plugin framework, developers can handle the pods with dependencies (e.g., the case that mentioned by me, some AI workload, etc) with straightforward logic, it is easy to maintain, and performance can be improved.

Updated "Alternatives" for the discussions in the readme, please help review them.

suigh commented:
@alculquicondor 
Thanks very much for your review and suggestions. 

For the solution of suspending a job, it works for sharing resources at job level, but it doesn't work for pod level sharing. E.g., if we can govern the resource sharing by stopping half of running pods in a job (the stopped pods will be re-created later), suspending the whole job is overkilled. 

Agree that `I'm not convinced that the scheduling queue sorting is gonna guarantee any type of resource sharing...`, only queue sorting cannot resolve this issue, we also need preemtption to preempt the pods from over-used users, then the pods from under-used users can run, and the KEP is to ensure the pods from under-used user can run first. Without this KEP, ping-pong preemption may happen as the pods from over-used users can be preempted and then get resource before other pods, and this step can be repeated.

For suggestion `But then we can have a PreFilter (and perhaps a Permit) plugin that would return Unschedulable if such pod would violate the resource sharing.`, I tried this one before, it is a little complex for developers, expecially when there are thousands of pods with different requests from differnt users, it is hard to say if a pod should run or not. For example, podA is from an over-used user, running it will violate the resource sharing, should we just mark it as `Unschedulable`? It is not true as if no other pods can run at all, podA should run to make full use of the cluster resources. It is hard to know the status of all other pods, even we can know, the performance will be impacted by going through all the pods, so it is hard to mark it as `Unschedulable`. With this KEP, we can control the sorting logic and always try the pods from under-used users first, and try other pods later, the above issue will not happen, the resources in the cluster can be fully used.

With this KEP and the scheduler plugin framework, developers can handle the pods with dependencies (e.g., the case that mentioned by me, some AI workload, etc) with straightforward logic, it is easy to maintain, and performance can be improved.

Updated "Alternatives" for the discussions in the readme, please help review them.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -396,3 +369,11 @@ Why should this KEP _not_ be implemented?
 An alternative is to enhance the current queue to meet some requests. The drawback
 is that we cannot handle all the potential requests as the effort is big, we should
 give the ball to the users, and they can do what they want.
```
I disagree with this statement.

I'm more in favor of keeping the interface small with clear use cases.

suigh replied:
The target of his KEP is to enable scheduler developers to implement their own logic to manage the order the pods. We can expose part of the interface for some known cases now, we will have to expose more interfaces in the future to handle new cases as developers' requests are un-preditable, this needs effort and maybe a new KEP. To expose the whole interface, we can handle the potentail requests by this KEP, needn't extra effort in the future.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -396,3 +369,11 @@ Why should this KEP _not_ be implemented?
+
+Another alternative is to make the pod that will break the resource sharing policy with
+`Unschedulable`. The drawback is that making such a decision needs the status of all other
+pods, the logic is complex, and the performance will be impacted.
```
In the proposal, don't you have to know the status of all pods during sorting? It looks like the latency is moving from one place to another.

suigh replied:
Oh, the `status` here is not clear. I update it to `The drawback is that making such a decision needs to know whether other
pods can run or not, ......`. To mark a pod as `Unschedulable`, the alternative needs to predict whether the competitor pods can run or not, this is complex.

In my proposal, the pods for different users will be put into different queues, and once the scheduler want to pop a pod, it will select the most under-used queue and pop one, the scheduler needn't care about other pods in other queues, logic is simple, there is no latency.

alculquicondor replied:
Can't you keep the same information about usage ratios in the PreFilter plugin? The latency should be similar.

alculquicondor reviewed (commented):
In general, I think the motivation is incomplete.

I would like to see how the interface would be used to satisfy the use cases you describe. The use cases themselves are still too high level. More detail will help. This would also inform better the "alternatives"

Then, your beta->GA criteria is basically to rewrite the whole interface and expose only "the necessary interface and structs". That's not the intent of graduation criterias. 
A graduating feature should have minimal variances from what is already released as alpha or beta. What you are proposing is that we rethink everything in the next release. I would like to know how this interface could look like now. Which brings us back to the point above: we need to understand better the use case.

Huang-Wei commented:
The key focus of this KEP is not to change the internal queue, or queue sorting plugin. Quite the contrary, that logic will be kept as is. The main idea is to give plugin developers an option to manage the whole queuing logic, including how to prioritize/pop/backoff/flush pods. It's all up to the plugin developers. And in terms of changes to the scheduler framework, it needs to make some functions public and tweak some parameters.

Another motivation of this KEP, IMHO, is to support multi-subqueue, which can be mapped to user-defined multi-tenancy model. If the plugin developers can implement their own scheduling queue, they're capable to manage pods for different tenants at ease. As we know, in the default scheduler, sorting in a global queue is somewhat awkward. I see some solutions are tagging pods to work it around, but building their own multi-subqueues can be more efficient and solve the problem fundamentally.

What @alculquicondor mentioned, managing the pods at creation time on the controller side is also a nice alternative. @suigh I think it'd be nice to compare the pros/cons, and polish the motivations/alternatives sections a bit. I'd expect to see at least compelling use cases that a custom scheduling queue can achieve (or achieve efficiently), while others cannot - which will make this KEP convicable.

suigh commented:
@alculquicondor @Huang-Wei 
Thanks very much for your review and suggestions, the readme is updated as following with your comments, would you please help review again?

1. Add following message at `Motivation` section to explain why we need to exponse the whole interface instead part of it.
```
This KEP is not used to change the internal queue, or queue sorting plugin, quite the contrary,
that logic will be kept as is. The main idea is to give scheduler developers an option to manage the whole
queuing logic, including how to prioritize/pop/backoff/flush pods. It's all up to the scheduler developers.
To ensure the developers can have full control of the queuing logic, the whole interface of the internal
queue will be exposed instead of part of it that can only meet above examples. By exposing the whole interface,
we needn't update the interface that need to be exposed again and again for coming demands, this can save
developers' effort as they needn't raise a KEP for such requests, this can also save community's effort.
```

2. Removed `Only necessary interface and structs are exposed to users` from ` Beta -> GA Graduation` section as this message may confuse the reader. What we will do is to expost the whole interface, not part of it.

3. Add more details in the `Alternatives` section to show what we do to resolve the multi-tenancy case, and what's the pros/cons when comparing with the possible alternatives.

alculquicondor commented on keps/sig-scheduling/2779-support-custom-queue-for-the-scheduler/README.md:
```diff
@@ -84,6 +84,14 @@ Given that the business needs may vary greatly, it'd be desirable to provide a r
 implement their custom queue implementation, while the upstream maintainers focus
 on keeping the core small and extensible, and thus only maintain the internal queue piece.
 
+This KEP is not used to change the internal queue, or queue sorting plugin, quite the contrary,
```
This should go as Non-Goal

suigh replied:
The above message is a little long for `Non-Goals`, how about updating the readme as following?

Add message `Enable scheduler developers to change the internal queue, or queue sorting plugin` in the `Non-Goals` section to show that it is not our goal.

Keep the orignal message in the `Motivation` section to let others know the details of our design.

alculquicondor commented:
In general, this change would be quite simple for us to maintain (just an exported interface)

However, I want to make sure you are not shooting yourselves in the foot by taking the challenge of building your own queue that might not solve the requirements you have.

Also, if you succeed in your multitenancy efforts, perhaps it's something that could be upstreamed. But as it is currently stated (you have a completely alternative implementation), it's gonna be harder to justify.

suigh commented:
For question `Can't you keep the same information about usage ratios in the PreFilter plugin? The latency should be similar.`

I guess your suggestion is to drop the pods from over-used user in PreFilter plugin, and only try to bind the pods from under-used user. This will cause latency when there are thousands of pods from over-used user. In the worst case, after we go through all the pods from under-used user and cannot get pod to run, we will have to re-check the dropped pods from over-used user.

suigh commented:
> In general, this change would be quite simple for us to maintain (just an exported interface)
> 
> However, I want to make sure you are not shooting yourselves in the foot by taking the challenge of building your own queue that might not solve the requirements you have.
> 
> Also, if you succeed in your multitenancy efforts, perhaps it's something that could be upstreamed. But as it is currently stated (you have a completely alternative implementation), it's gonna be harder to justify.

Thanks very much for your reminder. :)

I had implemented similar solution in my own repo, it can meet my multi-tenancy request. While it is hard for me to open source it as it is implemented on a non-kubernetes-official structure, and it is hard to sync up with the upstream. This is why this KEP was logged. Once this KEP was finished, I will add some codes at `kubernetes-sigs/scheduler-plugins` to show the details about the multi-tenancy solution.

alculquicondor commented:
Can @suigh  and @denkensk come to the next sig meeting?

suigh commented:
> Can @suigh and @denkensk come to the next sig meeting?

I can attend the meeting, and I'm not sure if Alex (@denkensk) is available or not. How long can we take to talk this ticket in the meeting? I didn't attend this meeting before.

> Another alternative: instead of exposing the entire SchedulingQueue, what about adding an interface just for `activeQ`?
Yes, it is a another solution that can help the multi-tenancy case. This solution is not used because:
1. scheduler developer cannot change the logic for backoff and unschedulable queues. E.g., in my project, backoff queue is not necessary at all, remove it can improve the performance.
2. developer don't have full control of the SchedulingQueue,  if a developer want to change something, she/he will have to talk with community, the effort is big.

denkensk commented:
> Can @suigh and @denkensk come to the next sig meeting?

Sure

suigh commented:
@alculquicondor, @denkensk,
I checked the meeting time, it is 10:00 AM PDT and 1:00 AM for my local time CST,  it is a little early.  Can we have the meeting in another time?   There is once a meeting at 5:00 PT, and 8:PM for CST, it is a good time, but it is cancelled. Can we reuse the time?

alculquicondor commented:
@Huang-Wei WDYT?

alculquicondor commented:
> in my project, backoff queue is not necessary at all, remove it can improve the performance.

That's the kind of things I would like to avoid. Removing the backoff can lead to starvation.

ahg-g commented:
I recently came back from vacation, I would like to take a look at this proposal, likely today or tomorrow.

suigh commented:
> > in my project, backoff queue is not necessary at all, remove it can improve the performance.
> 
> That's the kind of things I would like to avoid. Removing the backoff can lead to starvation.

Since developers have full control of the codes, the developers should handle all the related issues. If developers don't want to change the backoff queue, they can just leave it untouched, the result is same as only exposing the activeQ. We should try to remove the limitations on developers.

ahg-g commented:
The KEP is simply proposing a refactor to allow passing in different SchedulingQueue implementation on scheduler instantiation. 

I don't see this worthy of a KEP as it is proposed; also, as proposed, I don't think we will want to document this as a supported customization feature (like we do for plugins).

I think what would be worthy of support is a new SchedulingQueue interface that is abstract enough of the current internal implementation. For example, an interface that doesn't expose the details of Active/backoff/unschedulable queues etc.

Huang-Wei commented:
> @Huang-Wei WDYT?

If 5:00 PM PST works for both of you, I can setup a meeting next week.

Huang-Wei commented:
> I don't see this worthy of a KEP as it is proposed;

Probably this is a good way to move it forward, like what we did for some framework changes ([1](https://github.com/kubernetes/kubernetes/pull/103383), [2](https://github.com/kubernetes/kubernetes/pull/101394)) where the design doc was linked (but not a KEP).

Given that we don't have a clear plan for phase 2, a viable way is to claim minimum changes in a document that makes SchedulingQueue customizable, and evolve SchedulingQueue's abstraction in the long run. But if we go with a KEP, we may have to have concrete plans in the very beginning. @suigh WDYT?

ahg-g commented:
If other community members can't attend, then it is probably better to continue the discussion in this open forum.

suigh commented:
Making SchedulingQueue customizable with a document instead of KEP is good to me, can we go by this way? If yes, I will write a design document and update more details about motivation.

What we should in my mind:
1: Write a document about how to make the SchedulingQueue customizable instead of a KEP, and log a PR for the implementation. With this way, this feature is not a public one, but the developers can use this feature to implement their custom queues.
2: After the codes in step 1 are develiered, we will work on the SchedulingQueue abstraction, make it clean and simple for public usage.
3: Make this feature as a public one with a KEP when SchedulingQueue is clean/simple enough.

suigh commented:
@alculquicondor, @denkensk, @Huang-Wei, @ahg-g 

Thanks very much for your suggestions, help and patience on this KEP. 

Seems this is not the right time to expose the whole interface of the internal queue now, I will try simplifly the internal queue first as @ahg-g suggested, and maybe expose the refined interfacein the future. 

Close this KEP now, and thanks for your time again. :)

alculquicondor commented:
/close

Follow the discussion in kubernetes/kubernetes#102271 or open a new one
